### DISCUSSION TITLE
[@types/tar] is incompatible with new minipass types

### DISCUSSION THREAD

**Discussion by nex3** [DefinitelyTyped/DefinitelyTyped#60901] [Issues with a @types package]
[minipass 3.3.3](https://github.com/isaacs/minipass/releases/tag/v3.3.3) exposes its own typings, which aren't directly compatible with the `@types/minipass` typings that `@types/tar` was written for. This causes the following type error:

```
node_modules/@types/tar/index.d.ts:252:36 - error TS2709: Cannot use namespace 'MiniPass' as a type.
252 export interface ReadEntry extends MiniPass, HeaderProperties {
                                   ~~~~~~~~
Found 1 error in node_modules/@types/tar/index.d.ts:252
```

[a comment by an automated account is not shown]

**Reply by theaussiepom** [reply]
`import MiniPass from require('minipass');` imports the namespace rather than the type.
`export interface ReadEntry extends MiniPass, HeaderProperties {...` requires to extend the MiniPass type, not namespace.
We can import the type instead by `import MiniPass from 'minipass';`. However, this is considered bad practice and not allowed by DefinatelyTyped: [no-import-default-of-export-equals](https://github.com/microsoft/DefinitelyTyped-tools/blob/master/packages/dtslint/docs/no-import-default-of-export-equals.md)

By importing the namespace, we're actually trying to extend:
```
declare namespace MiniPass {
    interface Options {
        objectMode?: boolean | undefined;
        encoding?: string | null | undefined;
    }
}
```

What we're intending to extend is:
```
declare class MiniPass extends EventEmitter {
    readonly bufferLength: number;
    readonly flowing: boolean;
    readonly emittedEnd: boolean;
    readonly destroyed: boolean;
    ... ... ...
```

**Reply by theaussiepom** [reply]
We've spent quite some time on this and tried various solutions, mostly to no avail.
`export interface ReadEntry extends InstanceType<typeof MiniPass>, HeaderProperties {` appears to work at face value, but ultimately still calls type on the namespace.
Un-ideally copying the "MiniPass" type definition into "tar" _does_ work, identifying the class correctly (without the `InstanceType` hack above). This is because we're no longer importing it via a `require`. This further highlights the import is exposing the namespace rather than the class when used as a type.

**Reply by theaussiepom** [reply]
As a temporary solution, while still watching this thread, we're going to create a new git repo containing the bastardised tar type

**Reply by jcesarmobile** [reply]
did you create the new git repo?

**Reply by theaussiepom** [reply]
Please find a _temporary_ type published to [@theaussiepom/tar](https://www.npmjs.com/package/@theaussiepom/tar).
To install, run:
```
npm i @theaussiepom/tar
```
Once this issue is resolved, please revert to using @types/tar.

**Comment by danm** [comment]
temporary solution: depending on your package manager and your project, by adding `minipass@3.2.1` to your package.json, the dependencies will resolve a version of minipass which doesn't expose its own typing.

**Reply by joshspicer** [reply]
worked for me in the meantime 👍

**Comment by theaussiepom** [comment]
The issue is with the types, so addressing the types rather than the active logic would ensure we use the latest security and bug enhancements.

A temporary type has been published to [@theaussiepom/tar](https://www.npmjs.com/package/@theaussiepom/tar).
To install, run:

```
npm i @theaussiepom/tar
```
Once this issue is resolved, please revert to using @types/tar.

**Comment by jcesarmobile** [comment]
linking the issue for visibility
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/61073

**Comment by MysteryBlokHed** [comment]
I've made a PR that should fix the problem with minipass at isaacs/minipass#41, though it's been open for 16 days and hasn't been acknowledged yet.

**Comment by MysteryBlokHed** [comment]
isaacs/minipass#41 just got merged. Updating minipass to 3.3.5 should fix the typing problems
